driver-core/src/main/com/mongodb/internal/connection/DefaultAuthenticator.java

```diff
@@ -99,11 +98,27 @@ private Authenticator getLegacyDefaultAuthenticator(final ConnectionDescription
         }
     }
 
-    private BsonDocument createIsMasterCommand() {
-        BsonDocument isMasterCommandDocument = new BsonDocument("ismaster", new BsonInt32(1));
-        isMasterCommandDocument.append("saslSupportedMechs",
-                new BsonString(format("%s.%s", getMongoCredential().getSource(), getMongoCredential().getUserName())));
-        return isMasterCommandDocument;
+    protected Authenticator getAuthenticatorForIsMaster() {
+        AuthenticationMechanism mechanism = getMongoCredential().getAuthenticationMechanism();
+
+        if (mechanism == null) {
+            return new ScramShaAuthenticator(getMongoCredentialWithCache().withMechanism(SCRAM_SHA_256));
+        } else if (mechanism.equals(SCRAM_SHA_1) || mechanism.equals(SCRAM_SHA_256)) {
+            return new ScramShaAuthenticator(getMongoCredentialWithCache().withMechanism(mechanism));
+        } else if (mechanism.equals(MONGODB_X509)) {
+            return new X509Authenticator(getMongoCredentialWithCache().withMechanism(mechanism));
+        }
+        return null;
+    }
+
+    private void setSpeculativeAuthenticator(final ConnectionDescription connectionDescription) {
+        BsonArray saslSupportedMechanisms = connectionDescription.getSaslSupportedMechanisms();
```

Comment:
This is a behavioral change.  The call is assuming that speculativeAuthenticator is never null, yet this has a branch that leaves it null.  This is what makes sense to me:

        if (speculativeAuthenticator != null && ((SpeculativeAuthenticator) speculativeAuthenticator).getSpeculativeAuthenticateResponse() != null) {
            return;
        }

        if (connectionDescription.getSaslSupportedMechanisms() != null)  {
            BsonArray saslSupportedMechs = connectionDescription.getSaslSupportedMechanisms();
            AuthenticationMechanism mechanism = saslSupportedMechs.contains(DEFAULT_MECHANISM_NAME) ? SCRAM_SHA_256 : SCRAM_SHA_1;
            speculativeAuthenticator = new ScramShaAuthenticator(getMongoCredentialWithCache().withMechanism(mechanism));
        } else {
            speculativeAuthenticator = getLegacyDefaultAuthenticator(connectionDescription);
        }

I'm unclear why existing tests don't catch it.